go-controller/pkg/util/multi_network_test.go

```diff
@@ -1307,3 +1465,273 @@ func applyNADDefaults(nad *nadv1.NetworkAttachmentDefinition) *nadv1.NetworkAtta
+				Role:     ovntypes.NetworkRolePrimary,
+				Subnets:  "10.0.0.0/24",
+			},
+			hostSubnet: "10.0.0.0/24",
+			expectedIP: ovntest.MustParseIPNet("10.0.0.1/24"),
+		},
+		{
+			name: "Layer2 primary UDN with custom default gateway IP should return that custom IP",
+			netConf: &ovncnitypes.NetConf{
+				NetConf:           cnitypes.NetConf{Name: "l2-network"},
+				Topology:          ovntypes.Layer2Topology,
+				Role:              ovntypes.NetworkRolePrimary,
+				DefaultGatewayIPs: "10.0.0.5",
+				Subnets:           "10.0.0.0/24",
+			},
+			hostSubnet: "10.0.0.0/24",
+			expectedIP: ovntest.MustParseIPNet("10.0.0.5/24"),
+		},
+		{
+			name: "Layer2 primary UDN with infrastructure subnets should allocate from infrastructure subnet",
+			netConf: &ovncnitypes.NetConf{
+				NetConf:               cnitypes.NetConf{Name: "l2-network"},
+				Topology:              ovntypes.Layer2Topology,
+				Role:                  ovntypes.NetworkRolePrimary,
+				Subnets:               "10.0.0.0/24",
+				InfrastructureSubnets: "10.0.0.4/30",
+			},
+			hostSubnet: "10.0.0.0/24",
+			expectedIP: ovntest.MustParseIPNet("10.0.0.4/24"),
+		},
+		{
+			name: "Layer2 primary UDN with infrastructure subnets should allocate from infrastructure subnet",
+			netConf: &ovncnitypes.NetConf{
+				NetConf:               cnitypes.NetConf{Name: "l2-network"},
+				Topology:              ovntypes.Layer2Topology,
+				Role:                  ovntypes.NetworkRolePrimary,
+				Subnets:               "10.0.0.0/24",
+				InfrastructureSubnets: "10.0.0.0/30",
+			},
+			hostSubnet: "10.0.0.0/24",
+			expectedIP: ovntest.MustParseIPNet("10.0.0.1/24"),
+		},
+		{
+			name: "Layer2 primary UDN with infrastructure subnets should allocate from infrastructure subnet ignoring unusable IPs",
+			netConf: &ovncnitypes.NetConf{
+				NetConf:               cnitypes.NetConf{Name: "l2-network"},
+				Topology:              ovntypes.Layer2Topology,
+				Role:                  ovntypes.NetworkRolePrimary,
+				Subnets:               "10.0.0.0/24",
+				InfrastructureSubnets: "10.0.0.255/32, 10.0.0.9/32",
+			},
+			hostSubnet: "10.0.0.0/24",
+			expectedIP: ovntest.MustParseIPNet("10.0.0.9/24"),
+		},
+		{
+			name: "Layer2 primary UDN with infrastructure subnets IPv6",
+			netConf: &ovncnitypes.NetConf{
+				NetConf:  cnitypes.NetConf{Name: "l2-network"},
+				Topology: ovntypes.Layer2Topology,
+				Role:     ovntypes.NetworkRolePrimary,
+
+				Subnets:               "2001:db8::/64",
+				InfrastructureSubnets: "2001:db8::/126",
+			},
+			hostSubnet: "2001:db8::/64",
+			expectedIP: ovntest.MustParseIPNet("2001:db8::1/64"),
+		},
+		{
+			name: "Localnet topology should return traditional .1 address",
+			netConf: &ovncnitypes.NetConf{
+				NetConf:  cnitypes.NetConf{Name: "localnet-network"},
+				Topology: ovntypes.LocalnetTopology,
+			},
+			hostSubnet: "10.0.0.0/24",
+			expectedIP: ovntest.MustParseIPNet("10.0.0.1/24"),
+		},
+	}
+
+	for _, tc := range testCases {
```

Comment:
would it make sense to add a test for custom GW IP corner case ? i.e.  if the default management IP(.2) conflicts with the custom gateway IP, we should use the .1 address for the management IP.

Plus, would it make sense to add a test where we have *both* custom GW and custom mgmt port IPs ?